Center-align crosshairs, zoom range, and radar

ZenithMDC opened the pull request:
This fixes #273, ~~albeit with some room for improvement.~~

My original intention was to shift the post-scaled crosshair over by a resolution-dependent amount of pixels, in order to align it with the laser sight. However, I eventually noticed that the granularity of the rects was increased by 4x, once it was placed on the graphics display list (I'm not at all familiar with N64 technical details), which gave me enough precision to center it perfectly with the laser. Essentially, I solved this issue by creating a unique RDP opcode to handle the placement of the crosshairs. The original `gDPHudRectangle` macro was preserved, since the rect drawn around enemies, when aiming at them with a CMP150, was more accurately placed with it. In addition, the zoom range was also adjusted to align with the crosshair.

~~Some caveats to note:~~ **All caveats have been addressed.**
~~* With Crosshair Sway enabled, if the sway direction is up and/or left, it takes roughly one second for the crosshair to return to its origin. The original crosshair placement looks more correct, during this timeframe.~~
~~* The Skedar and Maian sights aren't perfectly aligned with the crosshair. The Maian one is much improved, but the Skedar one is actually slightly worse. I'm not even sure how, since I didn't modify how the Skedar sight was rendered at all.~~

<details><summary>Click here for some before and after screenshots, with a superimposed 32x32 grid to aid with visualizing the alignment. (OUTDATED)</summary>
<p>

Original crosshair (1x size):
![orig-crosshair](https://github.com/user-attachments/assets/b06ebd60-57a6-4ff1-8959-2ee8d18c71b7)
New crosshair (1x size):
![new-crosshair](https://github.com/user-attachments/assets/b831c564-655e-4fdb-b413-77a8f5a3e568)

Original classic sight:
![orig-classic](https://github.com/user-attachments/assets/0bd7755b-c964-48ff-8748-0c95aa91f202)
New classic sight:
![new-classic](https://github.com/user-attachments/assets/b94c79f3-8c6a-4d50-b788-cd63612d5e6c)

Original sniper zoom range (30x zoom)
![orig-sniper-zoom](https://github.com/user-attachments/assets/76271d79-af03-4b80-ad75-b783dbb743b6)
new sniper zoom range (30x zoom)
![new-sniper-zoom](https://github.com/user-attachments/assets/d7cd8e23-d2c0-44f6-abe8-0df03873cc94)

Original Maian sight:
![orig-maian-zoom](https://github.com/user-attachments/assets/de7d82c1-683f-44e5-8848-a805dd65d71f)
New Maian sight:
![new-maian-zoom](https://github.com/user-attachments/assets/1cf42673-0e49-43a3-95b5-f382c2395a97)

Original Skedar sight:
![orig-skedar-zoom](https://github.com/user-attachments/assets/70c04469-7a39-4d05-9d96-1a6eea884b32)
New Skedar sight:
![new-skedar-zoom](https://github.com/user-attachments/assets/23971bae-6122-4df8-b8f9-3b85b0475205)

Original DrugSpy crosshair:
![orig-drugspy](https://github.com/user-attachments/assets/c8c455c2-4e4a-4b6d-afe1-680677a4e2ed)
New DrugSpy crosshair:
![new-drugspy](https://github.com/user-attachments/assets/32f92a74-e896-4a2d-a0a4-53ae6db48895)

</p>
</details>

ZenithMDC committed:
Center-align crosshairs and zoom range

fgsfdsfgs commented:
I feel like this kind of thing might be better implemented as either an extended geometry mode or an extended state command that sets the subpixel offset. This way you would avoid changing all the hudrect calls, but also if the offset was customizable you could probably adjust the Skedar and Maian sights as well.

ZenithMDC commented:
To be honest, I just did the simplest possible things I could to get it to work. 😂 I couldn't figure out a way to get a resolution-dependent pixel offset for the HUD elements, which ultimately would have been ideal. Although, now that I think about it, a resolution-dependent implementation wouldn't work at smaller resolutions, due to quantization issues, whereas the current implementation works at any display resolution (`gfx_pc.cpp` always operates at a resolution of 1280x880). Internally, the game is treating it as if it's been drawn at 320x220. Any change to pixel offsets in `sight.c` or `bondview.c`, for instance, operate at that coarse resolution. I already implemented an extended command, `G_FILLRECT_CENTERED_WIDE_EXT`, to get this to work. Is this the same thing as an extended state command? I'm not sure. I'm open to making any changes at all, but I don't currently have an understanding of how to implement them using either an extended geometry mode or (maybe) extended state command.

fgsfdsfgs commented:
What I mean is instead of adding an extra FillRect variant you can do something like this:
```
gDPSetSubpixelOffsetEXT(gdl++, 2, 2);
gDPFillRectangle(gdl++, x + 2, y + 0, x + 7, y + 1);
```
and implement it by just doing this in the FillRect functions:
```
        ulx -= rdp.subpixel_ofs_x;
        lrx -= rdp.subpixel_ofs_x;
        uly -= rdp.subpixel_ofs_y;
        lry -= rdp.subpixel_ofs_y;
```
Also not sure what you mean by "gfx_pc is **always** 1280x880".

ZenithMDC commented:
Ah, I see. Thanks for the example. I didn't know about the rdp struct. I'll see to modifying my solution with your proposal.

> Also not sure what you mean by "gfx_pc is always 1280x880". Have you tested your solution at different resolutions?

For example, In `src/game/sight.c:1536`, from the debugger: ulx = x + 1 * SIGHT_SCALE = 161 (very close to 320 / 2).

The `gDPHudRectangleCentered` macro expands:
`_g1->words.w0 = _SHIFTL((ulx), 2, 22);` (only the line modifying ulx is shown)

and ulx is passed into the `_SHIFTL` macro. `_SHIFTL` modifies ulx to become 644, like so: `((((161) & ((0x01 << (22)) - 1)) << (2))) = 644`

which is the same as 161 * 4. This 644 ulx value is retrieved in `port/fast3d/gfx_pc.cpp`, once the `G_FILLRECT_CENTERED_WIDE_EXT` opcode is found in `gfx_run_dl()`. So, maybe to say "it's always 1280x880" isn't descriptive enough, but basically it's 4 times the native 320x220 resolution.

As for the second question: yes, my solution works at every resolution I've tried, including in 4:3 aspect. Here are a couple of screenshots.
640x480:
![640_480](https://github.com/user-attachments/assets/85bbafbc-7f0e-4db7-8627-bcfa4e17f207)
1280x720:
![1280x720](https://github.com/user-attachments/assets/a0819d17-2442-4faa-b11a-b8b8af6898f6)

ZenithMDC committed:
Replace centered rects with subpixel offsets

ZenithMDC commented:
Okay. I've made the requested changes. This is a much cleaner solution; thanks for suggesting it. If anything should be changed, let me know.

ZenithMDC commented:
I'm back. I just got the Maian sight perfectly center-aligned and I'm going to work on the Skedar sight next. I was putting it off because I figured it would be more difficult to do than it actually was. Luckily, no subpixel manipulation was required to align the Maian sight, anyway. I'll be committing the changes soon enough.

ZenithMDC committed:
Center Maian and Skedar tri sights

ZenithMDC committed:
Improve crosshair position accuracy by rounding

ZenithMDC committed:
Move subpixel offset code & fix radar UV map

The subpixel offset code has been moved into the draw_rectangle
function, so it can be applied to more than just the fill_rectangle
function.

The radar texture was displaced along its V component and has
been corrected. It has also been subpixel-adjusted to align with the
player's radar dot.

ZenithMDC commented:
UPDATE

Original radar: 
![orig-radar](https://github.com/user-attachments/assets/2b27edc5-9fe8-4496-a676-f94656896e4d)

Improved radar:
![new-radar](https://github.com/user-attachments/assets/7f33232d-4263-4d32-af89-9511a08bb3d9)

fgsfdsfgs commented:
Something that I don't like about this is rounding `crosspos`. That is used to calculate the actual aiming vector. You can't just round it in the sight drawing routine with no side effects, unfortunately. I'll change it so it uses a rounded copy of `crosspos` in sight.c instead, but that will not eliminate the "slow return to center" behavior.

ZenithMDC commented:
Whoops! Wasn't my intention to cause side effects with that change. Using a copy sounds like the best approach. As long as using the rounded copy of `crosspos` allows the crosshair to return to the center of the screen rapidly after being moved toward the up and/or left direction, then all is well. Thanks for reviewing & merging, btw.
